kv/txn_test.go

```diff
@@ -478,3 +478,172 @@ func TestTxnTimestampRegression(t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+// When the gap between two writes in a txn is bigger than 10 seconds,
+// lowWater in timestamp cache can increase to the timestamp of the
+// first write as side effect of read in another txn, but when the txn
+// committed, the write's timestamp is not increased. so another txn
+// two reads before and after the second write will get different
+// value.
+func TestTxnRepeatGetDifferentValue(t *testing.T) {
+	db, _, _, mClock, _, stopper, err := createTestDB()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer stopper.Stop()
+
+	keyA := proto.Key("a")
+	keyB := proto.Key("b")
+	ch := make(chan struct{})
+	// Use snapshot isolation
+	txnAOpts := &client.TransactionOptions{
+		Name:      "txnA",
+		Isolation: proto.SNAPSHOT,
+	}
+	txnBOpts := &client.TransactionOptions{
+		Name:      "txnB",
+		Isolation: proto.SNAPSHOT,
+	}
+	go func() {
+		err = db.RunTransaction(txnAOpts, func(txn *client.KV) error {
+			// Put transactional value.
+			if err := txn.Call(proto.Put, proto.PutArgs(keyA, []byte("value1")), &proto.PutResponse{}); err != nil {
+				return err
+			}
+			// Notify txnB do 1st get(b)
+			ch <- struct{}{}
+			// Wait for txnB notify us to put(b)
+			<-ch
+			// Write now to keyB
+			if err := txn.Call(proto.Put, proto.PutArgs(keyB, []byte("value2")), &proto.PutResponse{}); err != nil {
+				return err
+			}
+			return nil
+		})
+		// Notify txnB do 2nd get(b)
+		ch <- struct{}{}
+	}()
+
+	// Wait till txnA finish put(a)
+	<-ch
+	// Delay 12 seconds
+	mClock.Set(time.Duration(12 * time.Second).Nanoseconds())
```

Comment:
`time.Duration` is unnecessary here; `(12 * time.Second).Nanoseconds()` works. This should probably also be defined in terms of `minCacheWindow` instead of a separate hard-coded value.